src/publish-build/index.ts

```diff
@@ -0,0 +1,90 @@
+import * as core from '@actions/core';
+import {
+  S3Client,
+  PutObjectCommand,
+  PutObjectCommandOutput,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { promises as fs, createReadStream } from 'fs';
+import mime from 'mime';
+import * as path from 'path';
+
+const REQUIRED_ENV_VARS = [
+  'AWS_ACCESS_KEY_ID',
+  'AWS_SECRET_ACCESS_KEY',
+  'AWS_BUCKET',
+  'PROJECT',
+  'SOURCE_DIR',
+  'DESTINATION_DIR',
+  'COMMIT_HASH',
+];
+
+function validateEnvVars(): void {
+  for (const requiredVar of REQUIRED_ENV_VARS) {
+    if (!process.env[requiredVar]) {
+      const errMsg = `Required env variable missing: ${requiredVar}`;
+      core.error(errMsg);
+      throw new Error(errMsg);
+    }
+  }
+}
+
+async function getFileNames(dir: string): Promise<string[]> {
+  const dirents = await fs.readdir(dir, { withFileTypes: true });
+  const files = await Promise.all(
+    dirents.map(async dirent => {
+      const res = path.resolve(dir, dirent.name);
+      return dirent.isDirectory() ? getFileNames(res) : res;
+    }),
+  );
+  return Array.prototype.concat(...files);
```

Comment:
CMIIW, but it looks like this `Array.prototype.concat(...files);` call just gives us the original `files` array back.

Reply:
this will actually flatten the files out into a single array instead of nested. returning files gives ts error `Type '(string | string[])[]' is not assignable to type 'string[]'.`